qutip/qip/device/cqed.py

```diff
@@ -0,0 +1,517 @@
+        return ([r"$a^\dagger a$"] +
+                [r"$\sigma_x^%d$" % n for n in range(self.N)] +
+                [r"$\sigma_z^%d$" % n for n in range(self.N)] +
+                [r"$g_{%d}$" % (n) for n in range(self.N)])
+
+    def optimize_circuit(self, qc):
+        """
+        Function to take a quantum circuit/algorithm and convert it into the
+        optimal form/basis for the desired physical system.
+
+        Parameters
+        ----------
+        qc: :class:`qutip.QubitCircuit`
+            Takes the quantum circuit to be implemented.
+
+        Returns
+        --------
+        qc: :class:`qutip.QubitCircuit`
+            The circuit representation with elementary gates
+            that can be implemented in this model.
+        """
+        self.qc0 = qc
+        self.qc1 = self.qc0.resolve_gates(
+            basis=["SQRTISWAP", "ISWAP", "RX", "RZ"])
+        return self.qc1
+
+    def eliminate_auxillary_modes(self, U):
+        """
+        Eliminate the auxillary modes like the cavity modes in cqed.
+        """
+        return self.psi_proj.dag() * U * self.psi_proj
+
+    def load_circuit(self, qc):
+        """
+        Decompose a :class:`qutip.QubitCircuit` in to the control
+        amplitude generating the corresponding evolution.
+
+        Parameters
+        ----------
+        qc: :class:`qutip.QubitCircuit`
+            Takes the quantum circuit to be implemented.
+
+        Returns
+        -------
+        tlist: array-like
+            A NumPy array specifies the time of each coefficient
+
+        coeffs: array-like
+            A 2d NumPy array of the shape (len(ctrls), len(tlist)). Each
+            row corresponds to the control pulse sequence for
+            one Hamiltonian.
+        """
+        gates = self.optimize_circuit(qc).gates
+
+        dec = CQEDGateDecomposer(
+            self.N, self._paras, self.wq, self.Delta,
+            global_phase=0., num_ops=len(self.ctrls))
+        self.tlist, self.coeffs, self.global_phase = dec.decompose(gates)
+
+        # TODO The amplitude of the first control a.dag()*a
+        # was set to zero before I made this refactoring.
+        # It is probably due to the fact that
+        # it contributes only a constant (N) and can be neglected.
+        # but change the below line to np.ones leads to test error.
+        self.coeffs[0] = self._paras["w0"] * np.zeros((self.sx_u.shape[1]))
+        return self.tlist, self.coeffs
+
+
+class CQEDGateDecomposer(GateDecomposer):
+    """
+    Decompose a :class:`qutip.QubitCircuit` into
+    the pulse sequence for the processor.
+
+    Parameters
+    ----------
+    N: int
+        The number of qubits in the system.
+
+    paras: dict
+        A Python dictionary contains the name and the value of the parameters,
+        such as laser frequency, detuning etc.
+
+    wq: list of float
+        The frequency of the qubits calculated from
+        eps and delta for each qubit.
+
+    Delta: list of float
```

Comment:
They are left by whoever wrote this module first. It is common to refer to this variable as Delta when writing the Jaynes–Cummings Hamiltonian, so it might be more straightforward for those who are familiar with this model.